Add `concat_slice` and mark everything as unsafe

197g opened the pull request:
Chooses the very cautious option of erroring in all cases of a ZST
concatentation for now.

197g committed:
Make concatenation of raw slices public

Chooses the very cautious option of erroring in all cases of a ZST
concatentation.

197g committed:
Add unordered variant for slices

197g committed:
Test the behaviour of slice variants

oberien commented on src/lib.rs:
```diff
@@ -55,16 +55,64 @@ pub fn concat<'a>(a: &'a str, b: &'a str) -> Result<&'a str, Error> {
+/// than `isize::MAX`).
+///
+/// When T is a ZST then returns `Err(TooLong)` if the total length would overflow
+/// `usize` and `Err(NotAdjacent)` otherwise.
```
```suggestion
/// `usize` and `Err(NotAdjacent)` otherwise, because ZST-slices are
/// [weird](https://github.com/oberien/str-concat/issues/5) and
/// [hard](https://github.com/rust-lang/unsafe-code-guidelines/issues/93).
```

oberien commented on src/lib.rs:
```diff
@@ -55,16 +55,64 @@ pub fn concat<'a>(a: &'a str, b: &'a str) -> Result<&'a str, Error> {
+/// result is too long to be represented as a slice (size in bytes is larger
+/// than `isize::MAX`).
+///
+/// When T is a ZST then returns `Err(TooLong)` if the total length would overflow
```
```suggestion
/// When T is a zero-sized type (ZST) then returns `Err(TooLong)` if the total length would overflow
```

197g replied:
I've defined ZST just above that section in the same documentation so I'd rather be brief and `Strg-F` will help.

oberien commented on src/lib.rs:
````diff
@@ -55,16 +55,64 @@ pub fn concat<'a>(a: &'a str, b: &'a str) -> Result<&'a str, Error> {
+///
+/// ```rust
+/// # use str_concat::concat_slice;
+/// let s = b"0123456789";
````
I think it would be easier to read if you used `&[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]` to make the distinction between `concat` and `concat_slice` clearer.

197g replied:
It increases the line length considerably though. Did it anyways but it's unclear if it is clearer.

oberien commented on src/lib.rs:
````diff
@@ -55,16 +55,64 @@ pub fn concat<'a>(a: &'a str, b: &'a str) -> Result<&'a str, Error> {
+///
+/// ```rust
+/// # use str_concat::{concat_slice, Error};
+/// let s = b"0123456789";
````
~

oberien commented on src/lib.rs:
```diff
@@ -55,16 +55,64 @@ pub fn concat<'a>(a: &'a str, b: &'a str) -> Result<&'a str, Error> {
 
+    if mem::size_of::<T>() == 0 {
+        // Not safety critical but we check that the length is as expected.
+        if a_len.checked_add(b_len).is_none() {
```
I don't think we should make this check here. If the function is called with ZSTs, always return `NotAdjacent`. (also the check of `< isize::max_value()` elements would be required here instead of `< usize::max_value()`.)

197g replied:
`isize::max_value()` is the bound of the byte length. The byte length of a ZST slice is always `0`. The safety requirement is thus trivially fulfilled and the check here only checks if the result would be *correct*.

Also, it is possible to fit two ZST slices with length `usize::max_value()` into memory but we can surely not concatenate them. If we were to relax the requirements and allow concatenation of two ZST slices, ever, it might be beneficial to already return that error variant and have the check.

197g replied:
But the checks are in a different order than the non-zero checks so I see your point.

oberien commented on src/lib.rs:
```diff
@@ -55,16 +55,64 @@ pub fn concat<'a>(a: &'a str, b: &'a str) -> Result<&'a str, Error> {
+
+        // Never consider ZST slices adjacent. You could otherwise infinitely
+        // duplicate a non-zero length slice by concatenating it to itself.
+        return Err(Error::NotAdjacent)
```
Please include the reasoning here, or just the links to [this](https://github.com/oberien/str-concat/issues/5) and [this](https://github.com/rust-lang/unsafe-code-guidelines/issues/93).

oberien commented on src/lib.rs:
```diff
@@ -55,16 +55,64 @@ pub fn concat<'a>(a: &'a str, b: &'a str) -> Result<&'a str, Error> {
+    }
+
+    // `max_len <= isize::max_value()`
+    let max_len = isize::max_value() as usize / mem::size_of::<T>();
```
Shouldn't you use `mem::size_of::<T>() * {a,b}.len()` instead?

From the `pointer::offset` documentation:
> * The computed offset, **in bytes**, cannot overflow an isize.

197g replied:
`max_len` is the maximum `.len()` a slice of `[T]` is allowed to have, not the maximum of the passed in. So this computes the number of elements such that the size in bytes is still smaller than `isize::max_value()`. There is a note referring to this in the `SAFETY` section of `from_raw_parts` below.

oberien commented on src/lib.rs:
```diff
@@ -134,9 +184,50 @@ pub fn concat_unordered<'a>(a: &'a str, b: &'a str) -> Result<&'a str, Error> {
+pub fn concat_slice_unordered<'a, T>(a: &'a [T], b: &'a [T]) -> Result<&'a [T], Error> {
+    // add lengths to handle empty cases correctly
+    let a_ptr = a.as_ptr() as usize;
+    let a_end_ptr = a_ptr + a.len() * mem::size_of::<T>();
```
In theory this could overflow.

197g replied:
No, for the same reason as the `a_ptr + a.len()` in `concat_unordered` can not overflow. This one just computes the one-past-the-end-pointer of `a` but in bytes instead.

oberien replied:
Oh year, for some reason I read that as `a_ptr + b.len()`.

oberien reviewed (changes requested):
What do you think about combining `concat` and `concat_slice` like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0b3b8482ff7d6d1a2d79100a2fb3fc73)? I think it would make the usage more intuitive.

197g commented:
> What do you think about combining concat and concat_slice like this? I think it would make the usage more intuitive.

Adding a trait to the interface that can not be implemented seems to complicate the interface (it requires an additional indirection to find the valid uses). I'd either rename the methods to `concat` and `concat_str` or simply add the trait for `str` and `[T]` but not seal it.

```
trait Concat {
    fn concat(&self, other: &'_ Self) -> Result<&'_ Self, Error>;
}
```

197g committed:
Mark all primitive functions as unsafe

See issue #7, we must guarantee that slices are from the same
allocation. However, pointer equality testing can not reliably test this
so it must be provided by the caller.

Some actual safe interface is coming back at some point, currently the
proof obligation is still highly reduced.

197g committed:
Address review comments on documentation

197g committed:
Address remaining review comments

oberien commented on src/lib.rs:
```diff
@@ -134,56 +225,161 @@ pub fn concat_unordered<'a>(a: &'a str, b: &'a str) -> Result<&'a str, Error> {
+            assert_eq!(Err(Error::NotAdjacent), concat_slice(&s[..1], &s[2..]));
+            assert_eq!(Err(Error::NotAdjacent), concat_slice_unordered(&s[..1], &s[2..]));
+            assert_eq!(Err(Error::NotAdjacent), concat_slice(&s[2..], &s[..2]));
+        }
```
Could you please add a test-case for ZSTs?

oberien commented:
Could you please also add an example of `concat` to the Readme?
Edit: Could you please also wrap the current examples in `unsafe`?

197g committed:
Add negative test for ZSTs

197g committed:
Adjust documentation example to unsafe requirement
